bindings/java/Makefile

```diff
@@ -1,29 +1,89 @@
-.PHONY: gen_const clean jar all lib samples install
+all: jar lib samples
 
-all: gen_const
-	$(MAKE) -f Makefile.build all
+JC=javac
 
-lib:
-	$(MAKE) -f Makefile.build lib
+JAVA_HOME := $(shell readlink -f `which $(JC)` | sed "s:/bin/$(JC)::")
 
-samples:
-	$(MAKE) -f  Makefile.build samples
+JAVA_INC := $(shell realpath $(JAVA_HOME)/include)
 
-jar:
-	$(MAKE) -f Makefile.build jar
+JAVA_PLATFORM_INC := $(shell dirname `find $(JAVA_INC) -name jni_md.h`)
+
+UNICORN_INC=../../include
+
+SAMPLES := $(shell ls samples/*.java)
+TESTS := $(shell ls tests/*.java)
+SRC := $(shell ls unicorn/*.java)
+
+OS := $(shell uname)
+ifeq ($(OS),Darwin)
+   LIB_EXT=.dylib
+else ifeq ($(OS),Linux)
+   LIB_EXT=.so
+else
+   LIB_EXT=.dll
+endif
+
+CC=gcc
+CFLAGS=-fPIC
+LDFLAGS=-shared -fPIC
+LIBS=-lunicorn
+LIBDIR=-L../../
+INCS=-I$(JAVA_INC) -I$(JAVA_PLATFORM_INC) -I$(UNICORN_INC)
+
+CLASSPATH=./
+
+.SUFFIXES: .java .class
+
+tests/%.class: tests/%.java
+	$(JC) -Xlint:deprecation -classpath .:unicorn.jar:testdep/junit-4.13.2.jar $(JFLAGS) $<
+
+%.class: %.java
+	$(JC) -Xlint:deprecation -classpath .:unicorn.jar $(JFLAGS) $<
+
+OBJS=unicorn_Unicorn.o
+
+JARFILE=unicorn.jar
+
+%.o: %.c
+	$(CC) -c $(CFLAGS) $(INCS) $< -o $@
+
+unicorn_Unicorn.h: unicorn/Unicorn.java
+	$(JC) -h . $<
+
+unicorn_Unicorn.o: unicorn_Unicorn.c unicorn_Unicorn.h
+	$(CC) -O2 -Wall -Wextra -Wno-unused-parameter -c $(CFLAGS) $(INCS) $< -o $@
+
+libunicorn_java$(LIB_EXT): unicorn_Unicorn.o
+
+lib: libunicorn_java$(LIB_EXT) unicorn_Unicorn.h
+	$(CC) -o $< $(LDFLAGS) $(OBJS) $(LIBDIR) $(LIBS)
+
+samples: $(SAMPLES:.java=.class)
+tests: $(TESTS:.java=.class)
+jarfiles: $(SRC:.java=.class)
+
+jar: jarfiles
+	jar cf $(JARFILE) unicorn/*.class
+
+test: lib samples tests
+	java -Xcheck:jni -cp .:testdep/hamcrest-2.2.jar:testdep/junit-4.13.2.jar org.junit.runner.JUnitCore $(subst /,.,$(TESTS:.java=))
 
 install: lib jar
-	$(MAKE) -f Makefile.build install
+	cp libunicorn_java$(LIB_EXT) /usr/lib
+	cp $(JARFILE) /usr/share/java
```

Comment:
Is this the canonical location? Maybe `/usr/local` is more appropriate?

Reply:
I don't actually know! This is from the original Makefile. Honestly, the *right* solution would probably be to use Maven or Gradle to build a proper jar with the lib included, then put that jar on any classpath that needs Unicorn. If we want better packaging, I can look into doing that. (It might also be super cool to get this on Maven Central if there's interest, since that makes it easily available to any Java developer)

Reply:
I’m more than happy to publish this to maven central! But this involves a few more efforts which may be out of scope of this PR:

1. Setup the namespace on maven central which we could do this and get whatever token or credentials.
2. Setup CI to build this bindings regularly. Since uc2 release, we did rust and csharp for this which you may refer.
3. For each release, send the built release jar __together with__ prebuilt libraries to maven central. This ensures that anyone can use the bindings without installing unicorn globally.

Reply:
OK. At minimum, I'll look into adding a Maven POM to build the plugin automatically, and maybe look into hooking it up to the CI.

Reply:
I could help on tedious Github CI setup because I’m pretty familiar with it… You can just integrate it to maven and that’s a very good start already.

Reply:
Maven implementation is done in dfdc8e7e8ebc07e54eef08bba784601a129c8474. Didn't touch the CI stuff; I'll leave that to you :)